### PR TITLE
Add missing Entity + Projectile API

### PR DESCRIPTION
**Pull request by Owen1212055**
Adds a bunch of missing entity api, and some projectile api.

Please provide feedback on the javadocs, the shulker bullet api still needs to be tested and documented a bit better.

### CODE REVIEW

**Comment by Owen1212055** [comment]
Now adds __55__ new methods!

Many of these methods still need to be tested and documented better, will do overtime! 
### AbstractHorse
- Removed set/isEatingGrass deprecation & deprecated the upstream eat hay method. [See](https://canary.discord.com/channels/289587909051416579/925530366192779286/972682335810121738)
### Bat
- getTargetLocation
- setTargetLocation
Behavior: Tested and works, but appears to reset often so it needs to be set repeatedly if you want to force the bat somewhere.
### Bee
- setCropsGrownSincePollination
- getCropsGrownSincePollination
- setTicksSincePollination
- setTicksSincePollination
Behavior: Tested, using setTicksSincePollination with big numbers will cause the bee to try to go back to beehive. setCropsGrownSincePollination allows it so bees can only grow 10 crops while pollinated. 
### Chicken
- isChickenJockey
- setIsChickenJockey
- getEggLayTime
- setEggLayTime
Behavior: Tested, chicken jockey causes entity to drop more xp + lay egg time when set to 0 will cause chicken to lay egg.
### Endermite
- setLifetimeTicks
- getLifetimeTicks
Behavior: Tested and looks fine
### MushroomCow
- getStewEffectDuration
- setStewEffectDuration
- getStewEffectType
- setStewEffect
Behavior: Tested, looks good. However it should be noted that these values are reset when the mooshroom cow is interacted wtih.
### Phantom
- getAnchorLocation
- setAnchorLocation
Behavior: Tested, looks good.
### Polar Bear
- isStanding
- setStanding
### Ravager
- getAttackTicks
- setAttackTicks
- getStunnedTicks
- setStunnedTicks
- getRoarTicks
- setRoarTicks
Behavior: Tested, note animations do not play due to those using entity events. This however does allow you to do stuff like manually play the stun effect and more, but may not have an effect in certain cases.
### WanderingTrader
- getWanderingTowards
- setWanderingTowards
Behavior: Tested and is fine
### ExplosiveMinecart 
- setFuseTicks
- getFuseTicks
Behavior: Tested and there is no animation but does properly explode.
### HopperMinecart
- getPickupCooldown
- setPickupCooldown
Behavior: Bugged, see fix at https://github.com/PaperMC/Paper/pull/7804

### Llama
- getCaravanTail
- getCaravanHead
- hasCaravanTail
- leaveCaravan
- joinCaravan
- inCaravan

Projectile API:
### Projectile
- hasLeftShooter
- setHasLeftShooter
- hasBeenShot
- setHasBeenShot
### AbstractArrow
- setLifetimeTicks
- getLifetimeTicks
- getHitSound
- setHitSound
Behavior: Works fine for all
### ShulkerBullet
- getTargetDelta
- setTargetDelta
- getCurrentMovementDirection
- setCurrentMovementDirection
- getFlightSteps
- setFlightSteps
Behavior: Works, can get some pretty neat behavior
### Trident
- hasDealtDamage
- setHasDealtDamage
Behavior: Works

**Comment by Doc94** [comment]
Currently the Vex methods are already in upstream (with Vex#getBound and Vex#setBound).
PD: Sounds better Guarding than Bound :C ...

**Comment by Owen1212055** [comment]
Removed the vex api + added some more projectile api and polar bear methods

File: patches/api/0373-More-Projectile-API.patch [hunk 1]
```diff
@@ -134,6 +134,56 @@ index d1b37530319f6d37ee37f62080289c1e45848bc8..e94c7e279356c510f60508b26277d489
...
++     *
++     * @return has left shooter's hitbox
++     */
++    boolean hasLeftShooter();
++
++    /**
++     * Sets whether the projectile has left the
++     * hitbox of their shooter and can now hit entities.
++     *
++     * This is recalculated each tick if the projtile has a shooter.
```
Comment (molor) [inline patches/api/0373-More-Projectile-API.patch]: A mistake is above ~

File: patches/api/0312-Missing-Entity-Behavior-API.patch [hunk 2]
```diff
@@ -695,6 +695,24 @@ index 479f7a7c54c85cb685f56e60906650d1989c03ff..60267ee382de80fab86b440ff72a2455
 +     * @param standing whether the polar bear should be standing
 +     */
 +    void setStanding(boolean standing);
++
++    /**
++     * Gets when the polar bear will play its next
++     * warning sound when attacking an entity.
++     *
++     * This will tick down until it reaches 0, will it will
```
Comment (molor) [inline patches/api/0312-Missing-Entity-Behavior-API.patch]: Will it will? :D

Comment (Owen1212055) [reply patches/api/0312-Missing-Entity-Behavior-API.patch]: Lol thanks, alot of these javadocs are unfinished as I still need to test the exact behavior of some of these methods.
But feel free to point out any of these, 
thank you!

**Comment by Owen1212055** [comment]
Went ahead and tested all of these, looks good. However, the minecart hopper bug was discovered so that may have to be removed if it's not fixed. As I had a similar issue with the Rabbit moreCarrotTicks field, where mc just completely ignores it (https://bugs.mojang.com/browse/MC-249294), so I decided not to include api for that.

**Review by electronicboy — APPROVED** [review]

**Comment by Owen1212055** [comment]
Rebased

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 3]
```diff
@@ -182,6 +294,33 @@ index 94f3a8c4bf8cf14263d34d866db66728e98dfdb0..7937a0e082199554d3e8db1f9811be29
...
+--- a/src/main/java/org/bukkit/entity/Endermite.java
++++ b/src/main/java/org/bukkit/entity/Endermite.java
+@@ -23,4 +23,22 @@ public interface Endermite extends Monster {
+      */
+     @Deprecated
+     void setPlayerSpawned(boolean playerSpawned);
++    // Paper start
++    /**
++     * Sets the ticks for how long this endermite has been living.
++     * If this value 2400, the endermite will despawn.
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: "If this value 2400"

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 4]
```diff
@@ -182,6 +294,33 @@ index 94f3a8c4bf8cf14263d34d866db66728e98dfdb0..7937a0e082199554d3e8db1f9811be29
...
++     *
++     * @param ticks lifetime ticks
++     */
++    void setLifetimeTicks(int ticks);
++
++    /**
++     * Gets how long endermite enderman has been living for.
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: "endermite enderman"
So probably more a "Gets how long the endermite has been living for. This is only counted if the endermite has link persistence required bla bla"

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 5]
```diff
@@ -268,6 +407,104 @@ index d8eb2b5007091c25a14321cb389f3219d76ce452..0fc8a4fcc3ec2ce60bb095c31eb35333
...
+--- a/src/main/java/org/bukkit/entity/MushroomCow.java
++++ b/src/main/java/org/bukkit/entity/MushroomCow.java
+@@ -35,4 +35,38 @@ public interface MushroomCow extends Cow {
+          */
+         BROWN;
+     }
++    // Paper start
++
++    /**
++     * Gets how long that the effect that is
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: This could be improved in a few ways.

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 6]
```diff
@@ -268,6 +407,104 @@ index d8eb2b5007091c25a14321cb389f3219d76ce452..0fc8a4fcc3ec2ce60bb095c31eb35333
...
++     * applied to stew that is from this mushroom cow.
++     *
++     * @return duration of the effect (in ticks)
++     */
++    int getStewEffectDuration();
++
++    /**
++     * Sets how long that the effect that is
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: Same here.

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 7]
```diff
@@ -444,6 +708,73 @@ index 987f9b0866b213450b4de1310600161c8587a545..144fdcfd1f35b6346b672006905aedb8
...
+ /**
+  * Illager beast.
+  */
+-public interface Ravager extends Raider { }
++// Paper start - Missing Entity Behavior
++public interface Ravager extends Raider {
++
++    /**
++     * Gets for how many ticks the ravager is attacking for.
++     * During attacking, the ravager is cannot move.
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: "The ravanger is cannot move"

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 8]
```diff
@@ -444,6 +708,73 @@ index 987f9b0866b213450b4de1310600161c8587a545..144fdcfd1f35b6346b672006905aedb8
...
++     * @return ticks attacking or -1 if they are currently not attacking
++     */
++    int getAttackTicks();
++
++    /**
++     * Sets for how many ticks the ravager is attacking for.
++     * During attacking, the ravager is cannot move.
++     * This will tick down till it gets to -1, where it is no longer active.
++     *
++     * @param ticks ticks attacking or -1 if they are currently not attacking
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: or -1 if they should no longer be attacking.

The plugin would be passing the new value I presume.

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 9]
```diff
@@ -444,6 +708,73 @@ index 987f9b0866b213450b4de1310600161c8587a545..144fdcfd1f35b6346b672006905aedb8
...
++     * @return ticks stunned or -1 if they are currently not stunned
++     */
++    int getStunnedTicks();
++
++    /**
++     * Sets how many ticks the ravager is stunned for.
++     * The ravager cannot move or attack while stunned.
++     * At 0, this will cause the ravager to roar.
++     *
++     * @param ticks ticks stunned or -1 if they are currently not stunned
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: Same as above. Current state has nothing to do with the value.

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 10]
```diff
@@ -444,6 +708,73 @@ index 987f9b0866b213450b4de1310600161c8587a545..144fdcfd1f35b6346b672006905aedb8
...
++     */
++    int getRoarTicks();
++
++    /**
++     * Sets how many ticks the ravager is roaring for.
++     * While roaring, the ravager cannot move
++     * This will tick down till it gets to -1, where it is no longer active.
++     * If set to 11, this will play a sound and hurt nearby players.
++     *
++     * @param ticks ticks roaring or -1 if they are currently not roaring
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: Same as above. Current state has nothing to do with the value.

File: patches/api/0311-Missing-Entity-Behavior-API.patch [hunk 11]
```diff
@@ -563,6 +894,37 @@ index 627e3c1a96ae3331f5aa2dd7803dd2a31c7204be..3c447d2300c866ae605eeca97bd869f4
...
+     public boolean canDrinkMilk();
++
++    /**
++     * Gets the location that this wandering trader is currently
++     * wandering towards.
++     * <p>
++     * This will return null if the wandering trader has finished
++     * wandering towards the given location.
++     *
++     * @return wandering towards
```
Comment (lynxplay) [inline patches/api/0311-Missing-Entity-Behavior-API.patch]: the location the wandering trader is wandering towards

File: patches/api/0370-More-Projectile-API.patch [hunk 12]
```diff
@@ -5,6 +5,47 @@ Subject: [PATCH] More Projectile API
...
+diff --git a/src/main/java/org/bukkit/entity/AbstractArrow.java b/src/main/java/org/bukkit/entity/AbstractArrow.java
+index 225a24898acd25038ea2a8448f9f3b57643d3026..6b0358c5c159e784395d56caec6d33ddb0bd4278 100644
+--- a/src/main/java/org/bukkit/entity/AbstractArrow.java
++++ b/src/main/java/org/bukkit/entity/AbstractArrow.java
+@@ -151,6 +151,36 @@ public interface AbstractArrow extends Projectile {
+     @NotNull
+     org.bukkit.inventory.ItemStack getItemStack();
+ 
++    /**
++     * Sets the ticks for how long this arrow has been in the world for.
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch]: Sets the amount of ticks this arrow has been alive in the world.

File: patches/api/0370-More-Projectile-API.patch [hunk 13]
```diff
@@ -5,6 +5,47 @@ Subject: [PATCH] More Projectile API
...
++     * This is used to determine when the arrow should be automatically despawned.
++     *
++     * @param ticks lifetime ticks
++     */
++    void setLifetimeTicks(int ticks);
++
++    /**
++     * Gets the ticks for how long this arrow has been in the world for.
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch]: Same as above. I personally just thing `ticks for how long` is not a very pleasant thing to read.

File: patches/api/0370-More-Projectile-API.patch [hunk 14]
```diff
@@ -136,6 +177,120 @@ index d1b37530319f6d37ee37f62080289c1e45848bc8..e94c7e279356c510f60508b26277d489
...
++     * @return has been shot into the world
++     */
++    boolean hasBeenShot();
++
++    /**
++     * Sets whether the projectile has been
++     * shot into the world and has sent a projectile
++     * shot game event.
++     *
++     * Setting this to false will cause a game event
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch:222]: In the next tick.

File: patches/api/0370-More-Projectile-API.patch [hunk 15]
```diff
@@ -136,6 +177,120 @@ index d1b37530319f6d37ee37f62080289c1e45848bc8..e94c7e279356c510f60508b26277d489
...
+     void setTarget(@Nullable Entity target);
++
++    /**
++     * Gets where the shulker bullet should move towards.
++     * This is usually relative towards their target.
++     *
++     * @return target
++     */
++    @org.jetbrains.annotations.NotNull
++    org.bukkit.util.Vector getTargetDelta();
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch:246]: You might want to elaborate more on what a target delta here is, specifically in the shulker bullet case.
This changed on a tick base, the target delta setter should reference that.
It also isn't really the final destination, I think that would be good to reference here.

File: patches/api/0370-More-Projectile-API.patch [hunk 16]
```diff
@@ -136,6 +177,120 @@ index d1b37530319f6d37ee37f62080289c1e45848bc8..e94c7e279356c510f60508b26277d489
...
++     * that the bullet does not move in the same direction.
++     *
++     * Set to null to simply pick a random direction.
++     *
++     * @param movementDirection null or a direction
++     */
++    void setCurrentMovementDirection(@Nullable org.bukkit.block.BlockFace movementDirection);
++
++    /**
++     * Gets how many steps that the shulker bullet will try to move in
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch]: This can be reworded.

"Gets how many ticks this shulker bullet will at least remain in its current movement direction" or something along those lines.

This notabley is not related to the target delta, which changes on a tick to tick basis.

File: patches/api/0370-More-Projectile-API.patch [hunk 17]
```diff
@@ -136,6 +177,120 @@ index d1b37530319f6d37ee37f62080289c1e45848bc8..e94c7e279356c510f60508b26277d489
...
++     * the target delta.
++     *
++     * @return number of steps
++     */
++    int getFlightSteps();
++
++    /**
++     * Sets how many steps that the shulker bullet will try to move in
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch]: See above.

File: patches/api/0370-More-Projectile-API.patch [hunk 18]
```diff
@@ -179,3 +334,29 @@ index 7051e07b4e456aae0ec9e37808b59e5fa62a4027..225ac312613b9e8f3cf680819f2ebe35
...
++++ b/src/main/java/org/bukkit/entity/Trident.java
+@@ -38,5 +38,21 @@ public interface Trident extends AbstractArrow, ThrowableProjectile {
+      * @throws IllegalArgumentException if the loyalty level is lower than 0 or greater than 127
+      */
+     void setLoyaltyLevel(int loyaltyLevel);
++
++    /**
++     * Gets if this trident has dealt damage to an
++     * entity yet or has hit the floor.
++     *
```
Comment (lynxplay) [inline patches/api/0370-More-Projectile-API.patch:351]: I'd add a "If neither of these conditions are met, this method returns false" clarification here.

File: patches/server/0657-Missing-Entity-Behavior-API.patch [hunk 19]
```diff
@@ -55,6 +56,19 @@ index 04a119e6641898454253e2478bc1b4dff181b5ee..a8da601b8342aa6e4902b452eb588c76
...
+diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+index 5ffae9d3be22b5e78645da57a6bd0e7350749ef1..9aec9f80c564fa3ae03e445423d9e50afc45f837 100644
+--- a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
++++ b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+@@ -67,7 +67,7 @@ public class Llama extends AbstractChestedHorse implements RangedAttackMob {
+     @Nullable
+     private Llama caravanHead;
+     @Nullable
+-    private Llama caravanTail;
++    public Llama caravanTail; // Paper
```
Comment (lynxplay) [inline patches/server/0657-Missing-Entity-Behavior-API.patch:68]: I know we don't really care anymore, but any reason this is not an AT but so many others are ?

File: patches/server/0657-Missing-Entity-Behavior-API.patch [hunk 20]
```diff
@@ -368,6 +482,140 @@ index 7adda5c93e7c172ea0ba4a3f15828b5e54a283e7..fffaf4108b632ceabac4186d93b34ad0
...
++
++    // Paper start
++    @Override
++    public boolean inCaravan() {
++        return this.getHandle().inCaravan();
++    }
++
++    @Override
++    public void joinCaravan(@org.jetbrains.annotations.NotNull Llama llama) {
++        this.getHandle().joinCaravan(((CraftLlama) llama).getHandle());
```
Comment (lynxplay) [inline patches/server/0657-Missing-Entity-Behavior-API.patch:502]: I believe we would want a precondition check here (or a Validate, idk what the current flavour is we want). The annotation itself won't cut it at runtime.

**Review by lynxplay — CHANGES_REQUESTED** [review]
See comments.

File: patches/server/0657-Missing-Entity-Behavior-API.patch (hunk 19 above)
Comment (Owen1212055) [reply patches/server/0657-Missing-Entity-Behavior-API.patch:68]: This part of the pr is kinda newer, so, I just didn't bother changing that.

File: patches/server/0657-Missing-Entity-Behavior-API.patch (hunk 20 above)
Comment (Owen1212055) [reply patches/server/0657-Missing-Entity-Behavior-API.patch:502]: For what? Making sure that the llama doesn't try to join a caravan from ``null``? Tbh, in all my entity API I don't do any preconditions.

Comment (lynxplay) [reply patches/server/0657-Missing-Entity-Behavior-API.patch:502]: For a readable and understandable null pointer exception.
If someone passes null there, they get an NPE saying  "get handle cannot be called because llama is null"  or along the lines of that.

Comment (Owen1212055) [reply patches/server/0657-Missing-Entity-Behavior-API.patch:502]: This is a very similar PR which has something like this. https://github.com/PaperMC/Paper/pull/7089/files

Imo, null pointer should explain enough? But if others think that’s fine ig?
